manul/src/protocol/round_id.rs

```diff
@@ -0,0 +1,267 @@
+    pub id: RoundId,
+
+    /// Round IDs of the rounds that can finalize into this round.
+    pub parents: BTreeSet<RoundId>,
+
+    /// Round IDs of the other rounds that the parents of this round can finalize into.
+    ///
+    /// For example, a round can be followed either by a happy path round or an error round,
+    /// depending on the outcome of some checks during the finalization.
+    pub siblings: BTreeSet<RoundId>,
+
+    /// The round IDs of the rounds this round can finalize into.
+    ///
+    /// Returns an empty set if this round only finalizes into a result.
+    pub children: BTreeSet<RoundId>,
+
+    /// ``true`` if this round's [`Round::finalize`](`crate::protocol::Round::finalize`)
+    /// may return [`FinalizeOutcome::Result`](`crate::protocol::FinalizeOutcome::Result`).
+    pub may_produce_result: bool,
+}
+
+impl TransitionInfo {
+    /// Nest the round IDs under the given group. Used for combinators.
+    pub(crate) fn group_under(self, group: u8) -> Self {
+        Self {
+            id: self.id.group_under(group),
+            parents: self.parents.into_iter().map(|r| r.group_under(group)).collect(),
+            siblings: self.siblings.into_iter().map(|r| r.group_under(group)).collect(),
+            children: self.children.into_iter().map(|r| r.group_under(group)).collect(),
+            may_produce_result: self.may_produce_result,
+        }
+    }
+
+    /// Returns the set of round IDs that can be simultaneously active on other nodes
+    /// (not including the current round ID).
+    ///
+    /// This includes: the children rounds (if some nodes already finalized this round),
+    /// the parent rounds (if those nodes still haven't finalized while we already have),
+    /// and the sibling rounds (if some nodes went on a different path).
+    pub(crate) fn simultaneous_rounds(&self, followed_by_echo_round: bool) -> BTreeSet<RoundId> {
+        let mut result = if followed_by_echo_round {
+            BTreeSet::from([self.id.echo()])
+        } else {
+            self.parents.clone()
+        };
+        result.extend(self.siblings.iter().cloned());
+        result.extend(self.children.iter().cloned());
+        result
+    }
+
+    pub(crate) fn id(&self) -> RoundId {
+        self.id.clone()
+    }
+
+    /// Returns the corresponding transition info for the echo round following this one.
+    pub(crate) fn echo(self) -> Self {
+        Self {
+            id: self.id.echo(),
+            parents: [self.id.clone()].into(),
+            siblings: [].into(),
+            children: self.children,
+            may_produce_result: self.may_produce_result,
+        }
+    }
+
+    /// Creates a transition info for a non-terminating round in a linear sequence of rounds starting with 1.
+    ///
+    /// That is, if there are rounds 1, 2, 3, ..., N, where the N-th one returns the result,
+    /// this constructor can be used for rounds 1 to N-1.
+    pub fn regular(round_num: u8) -> Self {
```

Comment:
```suggestion
    pub fn new_regular(round_num: u8) -> Self {
```
Also: how about naming it `new_non_terminating` or `new_intermediate`?

Reply:
Renamed to `new_linear()` and `new_linear_terminating()`. Not entirely happy with the names still, but there's time until the next release.